### PR TITLE
 Expose rank_loss op Python API

### LINKED ISSUE #12117: fuild 0.14.0什么时候可以支持rank_cost ?

### PR DESCRIPTION
**Pull request by baiyfbupt**
resolve #12117

### CODE REVIEW

File: python/paddle/fluid/layers/nn.py [hunk 1]
```diff
@@ -5281,3 +5282,74 @@ def crop(x, shape=None, offsets=None, name=None):
         outputs={'Out': out},
         attrs=None if len(attrs) == 0 else attrs)
     return out
+
+
+def rank_loss(label, left, right, name=None):
+    """
+    **RankLoss layer for RankNet**
```
Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py]: RankLoss -> Rank loss

File: python/paddle/fluid/layers/nn.py [hunk 2]
```diff
@@ -5281,3 +5282,74 @@ def crop(x, shape=None, offsets=None, name=None):
...
+
+    (http://icml.cc/2015/wp-content/uploads/2015/06/icml_ranking.pdf).
+    RankNet is a pairwise ranking model with
```
Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py]: RankNet -> RankNet (http://icml.cc/2015/wp-content/uploads/2015/06/icml_ranking.pdf)

File: python/paddle/fluid/layers/nn.py [hunk 3]
```diff
@@ -5281,3 +5282,74 @@ def crop(x, shape=None, offsets=None, name=None):
...
+    one training sample consisting of a pair of doc A and B, and the label P
+    indicating that A is ranked higher than B or not:
+    
+    P = {0, 1} or {0, 0.5, 1}, where 0.5 means no information about the rank of
+    the input pair.
+    
+    The RankLoss layer takes three inputs: Left (o_i), Right (o_j) and Label
```
Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py]: The RankLoss layer -> This layer
Left (o_i), Right (o_j) and Label -> `left` (o_i), `right` (o_j) and `label`

**Review by qingqing01 — COMMENTED** [review]
@shanyi15 Please also help to review the doc. Thanks!

**Comment by shanyi15** [comment]
@reyoung please check the api reference，thanks

File: python/paddle/fluid/layers/nn.py (hunk 1 above)
Comment (baiyfbupt) [reply python/paddle/fluid/layers/nn.py]: Done

File: python/paddle/fluid/layers/nn.py (hunk 2 above)
Comment (baiyfbupt) [reply python/paddle/fluid/layers/nn.py]: Done

File: python/paddle/fluid/layers/nn.py (hunk 3 above)
Comment (baiyfbupt) [reply python/paddle/fluid/layers/nn.py]: Done

**Comment by shanyi15** [comment]
@baiyfbupt please don‘t merge before @skylarch review the grammar

File: python/paddle/fluid/layers/nn.py [hunk 4]
```diff
@@ -5281,3 +5282,74 @@ def crop(x, shape=None, offsets=None, name=None):
...
+
+
+def rank_loss(label, left, right, name=None):
+    """
+    **Rank loss layer for RankNet**
+
+    RankNet(http://icml.cc/2015/wp-content/uploads/2015/06/icml_ranking.pdf)
+    is a pairwise ranking model with one training sample consisting of a pair
+    of doc A and B, and the label P indicating that A is ranked higher than B
+    or not:
```
Comment (skylarch) [inline python/paddle/fluid/layers/nn.py]: RankNet(http://icml.cc/2015/wp-content/uploads/2015/06/icml_ranking.pdf)
is a pairwise ranking model with a training sample consisting of a pair
of documents, A and B. Label P indicates whether A is ranked higher than B
or not:

**Comment by baiyfbupt** [comment]
@shanyi15 Ok, Thanks!

File: python/paddle/fluid/layers/nn.py [hunk 5]
```diff
@@ -5281,3 +5282,74 @@ def crop(x, shape=None, offsets=None, name=None):
...
+    
+    P = {0, 1} or {0, 0.5, 1}, where 0.5 means no information about the rank of
+    the input pair.
```
Comment (skylarch) [inline python/paddle/fluid/layers/nn.py]: P = {0, 1} or {0, 0.5, 1}, where 0.5 means that there is no information about the rank of
the input pair.

File: python/paddle/fluid/layers/nn.py [hunk 6]
```diff
@@ -5281,3 +5282,74 @@ def crop(x, shape=None, offsets=None, name=None):
...
+    
+    This layer takes three inputs: left (o_i), right (o_j) and label
+    (P_{i,j}), which represent the output score of RankNet for the two docs and
+    the label respectively, and yields the rank loss C_{i,j} using the following
+    equation:
+
```
Comment (skylarch) [inline python/paddle/fluid/layers/nn.py]: Rank loss layer takes three inputs: left (o_i), right (o_j) and label (P_{i,j}). The inputs respectively represent RankNet's output scores for documents A and B and the value of label P. The following equation computes rank loss C_{i,j} from the inputs:

File: python/paddle/fluid/layers/nn.py [hunk 7]
```diff
@@ -5281,3 +5282,74 @@ def crop(x, shape=None, offsets=None, name=None):
...
+    the label respectively, and yields the rank loss C_{i,j} using the following
+    equation:
+    
+    $$
+      C_{i,j} = -\tilde{P_{ij}} * o_{i,j} + \log(1 + e^{o_{i,j}}) \\
+      o_{i,j} =  o_i - o_j  \\
+      \tilde{P_{i,j}} = \left \{0, 0.5, 1 \right \} \ or \ \left \{0, 1 \right \}
+    $$
+    
+    This layer can take batch inputs with size batch_size (batch_size >= 1).
```
Comment (skylarch) [inline python/paddle/fluid/layers/nn.py]: Rank loss layer takes batch inputs with size batch_size (batch_size >= 1).

File: python/paddle/fluid/layers/nn.py [hunk 8]
```diff
@@ -5281,3 +5282,74 @@ def crop(x, shape=None, offsets=None, name=None):
...
+    
+    Args:
+        label (Variable): The label indicating A ranked higher than B or not.
```
Comment (skylarch) [inline python/paddle/fluid/layers/nn.py]: label (Variable): indecates whether A is ranked higher than B or not.

File: python/paddle/fluid/layers/nn.py [hunk 9]
```diff
@@ -5281,3 +5282,74 @@ def crop(x, shape=None, offsets=None, name=None):
...
+        left (Variable): The output of RankNet for doc A.
+        right (Variable): The output of RankNet for doc B.
+        name(str|None): A name for this layer(optional). If set None, the layer
+                        will be named automatically.
+
```
Comment (skylarch) [inline python/paddle/fluid/layers/nn.py]: left (Variable): RankNet's output score for doc A.
right (Variable): RankNet's output score for doc B.
name(str|None): A name for this layer(optional). If set None, the layer will be named automatically.
这里”none“是否是name的default value？如果是default value，用户是否还需要亲自将name定义成”none“？”named automatically“是一种怎样的行为？layer是不是会被随意定义，即”randomly assigned“？
[可能的修改：name(str|None): a name for this layer(optional). If the variable remains to be the default value, "None", the layer will be randomly named.]

File: python/paddle/fluid/layers/nn.py [hunk 10]
```diff
@@ -5281,3 +5282,74 @@ def crop(x, shape=None, offsets=None, name=None):
...
+    Returns:
+        list: The output loss of RankLoss layer.
```
Comment (skylarch) [inline python/paddle/fluid/layers/nn.py]: list: the value of rank loss.

File: python/paddle/fluid/layers/nn.py [hunk 11]
```diff
@@ -5281,3 +5282,74 @@ def crop(x, shape=None, offsets=None, name=None):
...
+
+    Raises:
+        ValueError: If label or left or right is not a Variable.
```
Comment (skylarch) [inline python/paddle/fluid/layers/nn.py]: ValueError: any of label, left, and right is not a variable.

File: python/paddle/fluid/layers/nn.py (hunk 4 above)
Comment (baiyfbupt) [reply python/paddle/fluid/layers/nn.py]: Done, thanks!

File: python/paddle/fluid/layers/nn.py (hunk 5 above)
Comment (baiyfbupt) [reply python/paddle/fluid/layers/nn.py]: Done, thanks!

File: python/paddle/fluid/layers/nn.py (hunk 6 above)
Comment (baiyfbupt) [reply python/paddle/fluid/layers/nn.py]: Done, thanks!

File: python/paddle/fluid/layers/nn.py (hunk 7 above)
Comment (baiyfbupt) [reply python/paddle/fluid/layers/nn.py]: Done, thanks!

File: python/paddle/fluid/layers/nn.py (hunk 8 above)
Comment (baiyfbupt) [reply python/paddle/fluid/layers/nn.py]: Done, thanks!

File: python/paddle/fluid/layers/nn.py (hunk 10 above)
Comment (baiyfbupt) [reply python/paddle/fluid/layers/nn.py]: Done, thanks!

File: python/paddle/fluid/layers/nn.py (hunk 11 above)
Comment (baiyfbupt) [reply python/paddle/fluid/layers/nn.py]: Done, thanks!

File: python/paddle/fluid/layers/nn.py (hunk 9 above)
Comment (baiyfbupt) [reply python/paddle/fluid/layers/nn.py]: None是name的default value，
用户可以手动指定rank loss的name
也可以不指定，就自动设置为None, 会以`rank_loss_0.tmp_0`，`rank_loss_1.tmp_0`...的形式自动命名

**Review by qingqing01 — APPROVED** [review]
LGTM.

**Pull request merged**